### PR TITLE
[DPP-1356] Fix not logging some exceptions from streaming endpoints (B)

### PR DESCRIPTION
**Pull request by pbatko-da**
(An alternative fix: https://github.com/digital-asset/daml/pull/15749)

Sample error logged by the server:
```
18:11:30.818 [ForkJoinPool-4-worker-19] ERROR c.d.p.a.error.ErrorInterceptor - LEDGER_API_INTERNAL_ERROR(4,0): Unexpected or unknown exception occurred., context: {err-context: "{location=ErrorInterceptor.scala:73}"}
io.grpc.StatusRuntimeException: UNKNOWN
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at com.daml.platform.apiserver.error.ErrorInterceptor$$anon$1.close(ErrorInterceptor.scala:70)
	at io.grpc.PartialForwardingServerCall.close(PartialForwardingServerCall.java:48)
	at io.grpc.ForwardingServerCall.close(ForwardingServerCall.java:22)
	at io.grpc.ForwardingServerCall$SimpleForwardingServerCall.close(ForwardingServerCall.java:39)
	at com.daml.platform.apiserver.TruncatedStatusInterceptor$$anon$1.close(TruncatedStatusInterceptor.scala:19)
	at com.daml.platform.apiserver.MetricsInterceptor$TimedServerCall.close(MetricsInterceptor.scala:56)
	at io.grpc.PartialForwardingServerCall.close(PartialForwardingServerCall.java:48)
	at io.grpc.ForwardingServerCall.close(ForwardingServerCall.java:22)
	at io.grpc.ForwardingServerCall$SimpleForwardingServerCall.close(ForwardingServerCall.java:39)
	at com.daml.metrics.grpc.GrpcMetricsServerInterceptor$MetricsCall.close(GrpcMetricsServerInterceptor.scala:100)
	at io.grpc.stub.ServerCalls$ServerCallStreamObserverImpl.onError(ServerCalls.java:389)
	at com.daml.ledger.api.auth.Authorizer.$anonfun$authorizeWithReq$2(Authorizer.scala:283)
	at com.daml.ledger.api.auth.Authorizer.$anonfun$authorizeWithReq$2$adapted(Authorizer.scala:282)
	at scala.util.Success.fold(Try.scala:281)
	at com.daml.ledger.api.auth.Authorizer.$anonfun$authorizeWithReq$1(Authorizer.scala:285)
	at com.daml.ledger.api.auth.Authorizer.$anonfun$authorizeWithReq$1$adapted(Authorizer.scala:278)
	at com.daml.ledger.api.auth.services.TransactionServiceAuthorization.getTransactions(TransactionServiceAuthorization.scala:32)
	at com.daml.ledger.api.v1.transaction_service.TransactionServiceGrpc$TransactionService$$anon$1.invoke(TransactionServiceGrpc.scala:129)
	at com.daml.ledger.api.v1.transaction_service.TransactionServiceGrpc$TransactionService$$anon$1.invoke(TransactionServiceGrpc.scala:127)
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at io.grpc.Contexts$ContextualizedServerCallListener.onHalfClose(Contexts.java:86)
	at com.daml.ledger.api.auth.interceptor.AsyncForwardingListener.$anonfun$onHalfClose$1(AsyncForwardingListener.scala:37)
	at com.daml.ledger.api.auth.interceptor.AsyncForwardingListener.$anonfun$onHalfClose$1$adapted(AsyncForwardingListener.scala:37)
	at com.daml.ledger.api.auth.interceptor.AsyncForwardingListener.enqueueOrProcess(AsyncForwardingListener.scala:24)
	at com.daml.ledger.api.auth.interceptor.AsyncForwardingListener.onHalfClose(AsyncForwardingListener.scala:37)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at com.daml.platform.apiserver.error.ErrorListener.onHalfClose(ErrorInterceptor.scala:125)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:340)
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:866)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.RuntimeException: ALA123 11111111111
	at scala.sys.package$.error(package.scala:27)
	at com.daml.platform.server.api.services.grpc.GrpcTransactionService.getTransactionsSource(GrpcTransactionService.scala:47)
	at com.daml.ledger.api.v1.transaction_service.TransactionServiceAkkaGrpc.$anonfun$getTransactions$1(TransactionServiceAkkaGrpc.scala:11)
	at com.daml.grpc.adapter.server.akka.StreamingServiceLifecycleManagement.$anonfun$registerStream$1(StreamingServiceLifecycleManagement.scala:40)
	at com.daml.grpc.adapter.server.akka.StreamingServiceLifecycleManagement.ifNotClosed$1(StreamingServiceLifecycleManagement.scala:36)
	at com.daml.grpc.adapter.server.akka.StreamingServiceLifecycleManagement.registerStream(StreamingServiceLifecycleManagement.scala:38)
	at com.daml.grpc.adapter.server.akka.StreamingServiceLifecycleManagement.registerStream$(StreamingServiceLifecycleManagement.scala:27)
	at com.daml.platform.server.api.services.grpc.GrpcTransactionService.registerStream(GrpcTransactionService.scala:24)
	at com.daml.ledger.api.v1.transaction_service.TransactionServiceAkkaGrpc.getTransactions(TransactionServiceAkkaGrpc.scala:11)
	at com.daml.ledger.api.v1.transaction_service.TransactionServiceAkkaGrpc.getTransactions$(TransactionServiceAkkaGrpc.scala:7)
	at com.daml.platform.server.api.services.grpc.GrpcTransactionService.getTransactions(GrpcTransactionService.scala:24)
	at com.daml.ledger.api.auth.services.TransactionServiceAuthorization.$anonfun$getTransactions$1(TransactionServiceAuthorization.scala:31)
	at com.daml.ledger.api.auth.services.TransactionServiceAuthorization.$anonfun$getTransactions$1$adapted(TransactionServiceAuthorization.scala:31)
	at com.daml.ledger.api.auth.Authorizer.$anonfun$authorizeWithReq$3(Authorizer.scala:293)
	at com.daml.ledger.api.auth.Authorizer.$anonfun$authorizeWithReq$3$adapted(Authorizer.scala:285)
	... 35 common frames omitted
```

### CODE REVIEW

**Comment by skisel-da** [comment]
is it unit testable? there is at least `ErrorInterceptorSpec`

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala [hunk 1]
```diff
@@ -96,6 +101,12 @@ final class ErrorInterceptor extends ServerInterceptor {
     )
   }
 
+  private def isInternalErrorThrownFromScalapbGrpcCompleteObserver[RespT, ReqT](status: Status) = {
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala:104]:
```suggestion
  private def isUnsanitizedInternalOrUnknown[RespT, ReqT](status: Status) = {
```

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala:104]: How about handling the `UNKNOWN` in this method as well? After all, they convey the same thing albeit expressed differently depending on which lib generates them.

**Review by tudor-da — APPROVED** [review]
Thanks. Please add a unit test if possible.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala (hunk 1 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala:104]: My concern is that the original predicate is already somewhat unintuitive (as it tries to detect certain cases by making assumptions about scalapb implemention and our error categories) and now we're adding a conditional for a different second case. I tried to make the code reflect that we have those two distinct cases

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala:104]: ok, fine with not changing but please use at least a shorter name for the internal method e.g. `isUnsanitizedInternal`

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala [hunk 2]
```diff
@@ -54,24 +55,25 @@ final class ErrorInterceptor extends ServerInterceptor {
         * Knowing that Status.INTERNAL is used only by [[com.daml.error.ErrorCategory.SystemInternalAssumptionViolated]],
         * which is marked a security sensitive, we have the following heuristic: check whether gRPC status is Status.INTERNAL
         * and gRPC status description is not security sanitized.
+        * 2. Handling of Status.UNKNOWN:
+        * We do not have an error category that uses UNKNOWN so there is no risk of catching a legitimate Ledger API error code.
+        * A Status.UNKNOWN can arise when someone (us or a library):
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala:60]: I'm a bit concerned of some internal gRPC implementation logic using Status UNKNOWN for some specific purpose but don't have any examples of what could go wrong when we intercept and process it. 
Should be benign I guess

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala (hunk 1 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala]: Ok, done

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala:60]: **UNKNOWN** = ```Unknown error. For example, this error may be returned when a Status value received from another address space belongs to an error space that is not known in this address space. Also errors raised by APIs that do not return enough error information may be converted to this error.``` (see https://grpc.github.io/grpc/core/md_doc_statuscodes.html)

**Pull request merged**